google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java

```diff
@@ -1710,11 +1792,19 @@ private MultiplexedSession getOrNull() {
     @Override
     public MultiplexedSession get() {
       try {
-        MultiplexedSession multiplexedSession = super.get();
-        if (multiplexedSession != null) {
-          multiplexedSession.markBusy(span);
-          span.addAnnotation("Using Session", "sessionId", multiplexedSession.getName());
-          incrementNumSessionsInUse(true);
+        if (multiplexedSession == null) {
+          // access the settableApiFuture object before acquiring lock to avoid deadlock
```

Comment:
@olavloite this is something that I have added for the time being. But I am still debugging some more race conditions which is causing flaky failures in `SpannerGaxRetryTest`. I think there is still a scope to improve synchronization of the `SettableApiFuture` without causing deadlocks.

Reply:
Is it possible that multiple threads will call `MultiplexedSessionFuture#get()` (so this method) before it has been initialized? If so, then this would need its own lock.

Reply:
Yes, `getDatabaseClient(...).singleUse().read()` is completely non-blocking. So post app initialization multiple threads can get this object and issue a `next()` on the result set. When they do this there can be a race condition. Atleast some threads will get a `SettableApiFuture` which is not yet finished. And all of them will get blocked. 

In parallel another thread will issue the `CreateSession` call and will take a lock on `currentMultiplexedSessionReference` object in the `onSessesionReady()` callback function to initialize the session. This will unblock all the waiting threads. But I see the issue that in case the above thread are blocked on the uninitialized `SettableApiFuture` then that is already stale and these threads will eventually timeout.

However introducing a lock over here can introduce a deadlock? because some threads will acquire the lock on `currentMultiplexedSessionReference` and then `onSessesionReady` method will wait for this lock to be released. 

Do we need an additional waiter queue to mitigate this edge case?

Reply:
I modified the implementation a bit to use a double-checked locking mechanism to ensure that only one thread will initialize the object. That way, we effectively create a 'waiter' queue for the multiplexed session instance.